File: p4/deepspace/src/deepspace/SpaceCity.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author cristobalmer
 */
public class SpaceCity extends SpaceStation{
    
    private SpaceStation base;
    private ArrayList<SpaceStation> collaborators;
    
    public SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest){
        super(base);
        this.base = base;
        collaborators = new ArrayList<>(rest);
    }
    
    public ArrayList<SpaceStation> getCollaborators(){
        return collaborators;
    }
    
    @Override
    public float fire(){
        float factor = base.fire();
        for(int i=0;i<collaborators.size();i++){
            factor += collaborators.get(i).fire();
        }
        return factor;
    }
    
    @Override
    public float protection(){
        float factor = base.protection();
        for(int i=0;i<collaborators.size();i++){
            factor += collaborators.get(i).protection();
        }
        return factor;
    }
    
    @Override
    public Transformation setLoot(Loot loot){
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }
    
    @Override
    public SpaceCityToUI getUIversion() {
        return new SpaceCityToUI(this);
    }
}

File: p4/Deepspace/SpaceCity.rb
require_relative 'SpaceCityToUI'
require_relative 'SpaceStation'

module Deepspace
    class SpaceCity < SpaceStation
        
        def initialize(base, rest)
            super(base.name(),SuppliesPackage.new(base.ammoPower(),base.fuelUnits(),base.shieldPower()))
            @collaborators = rest
            @base = base
            @nMedals = base.nMedals
				
	    @hangar = base.hangar
            @pendingDamage = base.pendingDamage
	    @shieldBoosters = base.shieldBoosters 
	    @weapons = base.weapons 
        end
        attr_reader :collaborators
        
        def fire()
        	potencia = super
        	@collaborators.each do |station|
        		potencia += station.fire()
        	end
        	
        	return potencia
        end
        
        def protection()
        	potencia = super
        	@collaborators.each do |station|
        		potencia += station.protection()
        	end
        	
        	return potencia
        end
        
        def setLoot(loot)
        	super(loot)
        	return Transformation::NOTRANSFORM
        end
        
        def getUIversion()
		Deepspace::SpaceCityToUI.new(self)
	end
    end
    #station = SpaceStation.new("p",SuppliesPackage.new(1,2,1))
    #hangar = Hangar.new(3)
    #hangar.addWeapon(Weapon.new("Láser infinito",WeaponType::LASER,100))
    #station.receiveHangar(hangar)
    #station.mountWeapon(0)
    #puts station.fire()
    #puts station.protection()
    #station2 = SpaceStation.new("p2",SuppliesPackage.new(2,2,2))
    #puts station2.fire()
    #puts station2.protection()
    #station3 = SpaceStation.new("p3",SuppliesPackage.new(3,2,3))
    #puts station3.fire()
    #puts station3.protection()
    #rest = [station2,station3]
    #puts rest
    #pwsp = SpaceCity.new(station,rest)
    #puts pwsp.fire()
    #puts pwsp.protection()
end
